### ISSUE TITLE
Sequence.on_epoch_end signature is different from Callback.on_epoch_end

### ISSUE THREAD

**Issue by jatir** [keras-team/keras#7851]
- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [ ] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [ ] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

Sequence.on_epoch_end signature is different from Callback.on_epoch_end.  As they conflict, an object can't implement both.  Either Sequence should be changed to have the same signature as Callback, or the method should be renamed so it doesn't conflict.  

I don't mind making the change, but I need to know which option is preferable.

**Comment by Dref360** [comment]
What is your use case for this? Right now, Sequences do not have access to the epoch logs like a callback. 

renaming would be a breaking change, but Sequences are not really used right now.

**Comment by jatir** [comment]
Was using the same object (with different data) for fit_generator training and validation generators.  fit_generator was only accepting a Sequence for the validation_data generator, although I can't reproduce this anymore.

I needed to shuffle the data on epoch start, rather than epoch end, as the Sequence api allowed.

If Sequences aren't the recommended approach, I can change things to regular generators to work around this issue.

**Comment by rohit-gupta** [comment]
Hi @Dref360 @jatir I am also trying to use the Sequence API and was wondering where I could find more detailed documentations for it ?

**Comment by Dref360** [comment]
I'll start working on a guide, but basically, you extend keras.utils.Sequence. Implements the required method (__len__, __get_item__ and on_epoch_end). When you do `youseq[i]` this should return the batch number i. That's pretty much it. 

@jatir. What I would recommend is to stick to on_epoch_end. The timeframe between the Sequence's `on_epoch_end` is not the same as the `on_epoch_end` of callbacks. There may be `queue_size` batches between the two.
